### PR TITLE
expr,sql: fix handling of JSON nulls

### LINKED ISSUE #25899: sqlsmith: internal error: column 3 is not of expected type ColumnType { scalar_type: Jsonb, nullable: true }: Map({"89": Null})
### What version of Materialize are you using?

f9ea5c54061d38e9fccb700773742641ab454a05

### What is the issue?

```
materialize=> SELECT NULL AS c2, pg_catalog.jsonb_object_agg(subq_2.c2, mz_internal.mz_normalize_schema_name(NULL)) AS c3 FROM (SELECT subq_1.c9 AS c2 FROM (SELECT subq_0.c0 AS c9 FROM LATERAL (SELECT 89 AS c0) AS subq_0) AS subq_1) AS subq_2;
ERROR:  internal error: column 1 is not of expected type ColumnType { scalar_type: Jsonb, nullable: true }: Map({"89": Null})
```

Logs:
```
environmentd: 2024-03-08T22:49:15.428893Z  WARN mz_transform::typecheck: TYPE ERROR IN TRANSIENT QUERY:
In the MIR term:
Constant
  - (█, █)


bad constant row
      got (null, {89: null})
expected row of type (String?, Jsonb)

Constant
  - (█, █)

environmentd: 2024-03-08T22:49:15.428896Z ERROR mz_transform::typecheck: type error in MIR optimization (details in warning; see 'Type error omnibus' issue #19101 <https://github.com/MaterializeInc/materialize/issues/19101>)
```

I'm scared that this might be a correctness issue. We had a similar issue in https://github.com/MaterializeInc/materialize/issues/23015 @aalexandrov are you the right assignee?

ci-regexp: is not of expected type ColumnType { scalar_type: Jsonb,

### PR DESCRIPTION
**Pull request by benesch**
Fix to_jsonb's handling of NULL input values. To match PostgreSQL, to_jsonb should return a SQL NULL whenever its input is NULL. Previously we were returning a JSON null.

Also, fix a glitch where converting a list of arrays to JSON could panic if one of the arrays in the list was NULL.

Supersedes #25934.

Close #25934.
Fix #25899.

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Adjust `to_jsonb`'s handling of `NULL`s to match PostgreSQL. Specifically `to_jsonb` now returns `NULL` when its input is `NULL`, rather than returning the JSON `null` value.
  - Fix a panic when calling `to_jsonb` on a list containing NULL array values.

### CODE REVIEW

**Comment by def-** [comment]
I'll fix the SLT failures and run nightly on this (and SQLsmith locally)

File: test/sqlsmith/mzcompose.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 from materialize.sqlsmith import known_errors
 
 TOTAL_MEMORY = 8
-NUM_SERVERS = 2
+NUM_SERVERS = 8
```
Comment (mjibson) [inline test/sqlsmith/mzcompose.py:26]: Is this supposed to be here?

**Review by mjibson — APPROVED** [review]

File: test/sqlsmith/mzcompose.py (hunk 1 above)
Comment (def-) [reply test/sqlsmith/mzcompose.py]: Sorry, no. That was for my local test run. Fixed

**Review by ggevay — APPROVED** [review]

**Review by def- — APPROVED** [review]
Seems good, ran into the expected error a few times with SQLsmith, so seems to work as expected. No further issues found.

File: src/sql/src/func.rs [hunk 2]
```diff
@@ -3069,8 +3069,12 @@ pub static PG_CATALOG_BUILTINS: Lazy<BTreeMap<&'static str, Func>> = Lazy::new(|
                     _ => val,
                 };
 
+                let json_null = HirScalarExpr::literal(Datum::JsonNull, ScalarType::Jsonb);
                 let key = typeconv::to_string(ecx, key);
-                let val = typeconv::to_jsonb(ecx, val);
+                let val = HirScalarExpr::CallVariadic {
+                    func: VariadicFunc::Coalesce,
+                    exprs: vec![typeconv::to_jsonb(ecx, val), json_null],
+                };
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/sql/src/func.rs:3077]: Good point. It's a slightly different reason though. I put up https://github.com/MaterializeInc/materialize/pull/25951 with the explanation.

**Comment by benesch** [comment]
Thanks very much for the review, @ggevay, and for the testing, @def-!

**Pull request merged**